src/messaging/protocol.ts

```diff
@@ -69,3 +70,47 @@ export function isRemoteProcedureCallRequest(
   // eslint-disable-next-line @typescript-eslint/no-explicit-any -- This is a type guard function and it uses ?.
   return typeof (message as any)?.type === "string";
 }
+
+export type MessageListener = (
+  request: Message,
+  sender: Runtime.MessageSender
+) => Promise<unknown> | void;
+
+export function allowSender(sender: Runtime.MessageSender): boolean {
+  return sender.id === browser.runtime.id;
+}
+
+// https://www.typescriptlang.org/docs/handbook/2/conditional-types.html
+export type MessageTypeOf<M> = M extends Message<infer T> ? T : never;
+
+export class HandlerMap<
+  ActionType extends string,
+  Action extends Message<ActionType>,
+  Handler extends (
+    request: Action,
+    sender: Runtime.MessageSender
+  ) => Promise<unknown> | void
+> extends Map<ActionType, Handler> {
+  // The typing isn't quite right here. The actionType/action correspondence doesn't get enforced at the
+  // call-site because the actionType is just inferred to be string and not the typeof the constant being
+  // passed in.
+  set(actionType: ActionType, value: Handler): this {
+    if (this.has(actionType)) {
+      throw new Error(`Handler for ${actionType} already defined`);
+    }
+    super.set(actionType, value);
+    return this;
+  }
+
+  asListener(): MessageListener {
+    return (request: Action, sender: Runtime.MessageSender) => {
```

Comment:
The actual request is of type `any`:

<img width="549" alt="Screen Shot" src="https://user-images.githubusercontent.com/1402241/125039730-0315c380-e0c1-11eb-948e-2a01c11b8a45.png">

So this line is acting as an assertion. We probably need a type guard like:
```suggestion
    return (request: unknown, sender: Runtime.MessageSender) => {
    	if (!isAction(request)) {
    		return;
    	}
```

Reply:
> So this line is acting as an assertion

We get the type check for free. If the "type" property isn't defined on the object, no handler will match